Revamped Website 2019

gmanlan opened the pull request.

jeffin143 commented:
Is there any way we can view the website locally ?

rcurtin commented:
@jeffin143: I've got a version of the website built at http://mlpack2.ratml.org; you can take a look at it there. :+1:

conradsnicta commented:
@gmanlan @rcurtin  That red ribbon on the top-right ("GitHub") is quite distracting.

It's also a bad idea from a branding / marketing point of view -- the mlpack site is supposed to be about mlpack, not an advertisement for GitHub.

gmanlan commented:
Hey @conradsnicta, thanks for the feedback. I don't think it's a big deal, as it's not promoting GitHub per se but linking to the mlpack repo (besides, today we already have a link to the GitHub repo anyways so...). Many libraries do that -> i.e. https://scikit-learn.org/stable/

Nevertheless, I will go ahead and remove it.
Thanks

gmanlan commented:
Ok, so I removed the github ribbon, fixed some paths to work with the deployment strategy and fixed merge conflicts. @rcurtin if you have some time to re-deploy the test stage of the site, we should be able to see the changes.

rcurtin commented:
Rebuilt and deployed. :+1:

gmanlan commented:
> Rebuilt and deployed. 👍

@rcurtin : did you re-build the Doxygen doc as well? I see the logo missing in the doc header, as well as an extra (removed already) github link. Maybe because the doc is using a previous header?

rcurtin commented:
@gmanlan good point, I missed that part.  I'm rebuilding the doxygen versions now, but at least the binding documentation has the headers correct now.  Sorry about that.

rcurtin commented on _includes/footer.html:
```diff
@@ -1 +1 @@
-<footer><span></span></footer>
+<footer><span>mlpack.org</span></footer>
```
Do you think maybe we should put the mlpack logo here also?  The `mlpack.org` seems a bit plain to me.  Maybe we want to right-justify it or something too?

gmanlan replied:
Checkout the footer of this page, would that be a good txt footer? https://scikit-learn.org/stable/install.html

© 200? - 2019, mlpack developers (BSD License).

Logo would be redundant I guess...

rcurtin replied:
Ah yeah, that could work too. mlpack's first line of code was written in 2007 also, just like scikit. :)

rcurtin commented on _src/doxygen/doxygen.css:
```diff
@@ -193,8 +203,9 @@ div.fragment {
-        color: #ddd;
+	font-size: 85%;
+	font-weight: 900;
+	color: #616161;
```
This seems a little light on a background of #dedede, maybe we should make this darker?  Example:
http://mlpack2.ratml.org/doc/mlpack-3.1.1/doxygen/timer.html

Also personally I think the `font-weight` makes it too thick, just my opinion.  Although I guess I can see it either way.  So that's certainly not a strong opinion.

gmanlan replied:
I'm not sure what you refer to, do you have an example to see where the problem is? (tracking doxygen css is painful).

rcurtin replied:
Oh, sorry, I just meant the code snippets there.  For instance the first one on the page:

```
void Timer::Start(const char* name);
void Timer::Stop(const char* name);
timeval Timer::Get(const char* name);
```

I think the gray color is a bit light, at least on my monitor. :)

gmanlan replied:
Fixed!

rcurtin commented on _src/doxygen/doxygen.css:
```diff
@@ -658,31 +664,28 @@ td.memItemLeft div.content, td.memTemplItemLeft div.content {
 
 .overload {
-        font-family: "courier new",courier,monospace;
+    font-family: "courier new",courier,monospace;
```
Hmm, should we use `Roboto` here?  Looks like I might have overlooked that in the past.

gmanlan replied:
Done, but we should see how it looks like once you rebuild doxygen.

rcurtin commented on _src/doxygen/doxygen.css:
```diff
@@ -1802,7 +1801,7 @@ td.memTemplItemLeft div.hover
 {
-  background: #202020 !important;
+  
 }
```
We can just remove this whole block now I guess. :)

gmanlan replied:
Done

rcurtin commented on docs.md:
```diff
@@ -1,6 +1,37 @@
-order: 2
+order: 3
 ---
+<div class="page-title-header">Documentation</div>
```
Maybe should we just use the H1 class for this here?

```
# Documentation
```

plus the right CSS for `h1` that's currently in `page-title-header`?  It might simplify things a bit.

gmanlan replied:
Same comment as before, the class is a rich CSS - not just font size.

rcurtin commented on docs.md:
```diff
@@ -1,6 +1,37 @@
 ---
+<div class="page-title-header">Documentation</div>
+
+<h2>Overview</h2>
```
```suggestion
## Overview
```

These should be equivalent, I think.

gmanlan replied:
Correct, fixed.

rcurtin commented on getstarted.md:
```diff
@@ -0,0 +1,43 @@
+
+Here is a summary of the currently available distribution options you can use depending on your needs:
+
+<h2>Python</h2>
```
In this file also I think we can replace all the `<hX>` with Markdown headings, and the links with Markdown links.

gmanlan replied:
Done. Could you please review all the links and update them accordingly? When I started this project we were referring to mlpack 3.1 and new versions have been released. Also, I believe we should be using a static link instead of versioned (i.e. stable).

rcurtin replied:
I didn't do my homework, which was to figure out how to push the current state of mlpack.org back to the repository.  I'll do that now, but I know it will cause merge conflicts on this branch.  I can work them out too and push the changes / provide a patch, or just work it out during the actual merge of the PR; up to you.

gmanlan replied:
I think it will be faster if you solve the conflicts and/or manually update the files to use the correct links. Then moving forward we can make sure it's a seamless process.

zoq commented:
> My biggest subjective comment is that I think we should try and think of a way to relate the imagery/design on the homepage to something more "machine learning" or "math"-y. I guess probably the reason for that comment is mostly the hero image, but maybe there are some other things we could do also, not sure. I'll think about some other ideas, but I don't know what I'll be able to come up with.

I have a couple of ideas that could work here, I'll sketch something up and post an update later today.

zoq commented:
I tested out different ideas and combined some of them that you can see in the example images below. Also, I haven't touched anything on the subpages, I think they are great so I would probably only change the color-scheme.

In the last meeting, we had a discussion, about how cool notebooks are, so I tried to pick that idea up for the frontpage. So the editor you see isn't a gif or video but a simple editor (ace) that acts as a frontend to run code on a remote machine (in a docker container or something). The output is shown in the white box below. Note the output doesn't match with the example code. So a user could use that as a simple and quick way to run mlpack without installing anything.

I thought the focus wasn't on the actual code example with the reddish polygon background, so I tried to find a more nutural color that could work, unfortunately, I couldn't find a reddish color so I went with blue.

I've updated some of the sections below as well, like the citation section, which now shows more than one paper and added links to the BibTeX file, google scholar and the pdf file; not sure we should list more than one paper.

Haven't really changed the cards; slightly decreased the corner-radius and adapted the color scheme.

There is also a new benchmarks section, I'm not sure if I like it or not; the current minimal approach might be better.

As Ryan pointed out we should add a NumFOCUS section as well, so I did that and just copied the text from the Github page.


Interested to hear what you think about some of the ideas.

![alt text](https://kurg.org/media/mlpack-www-1.jpg)

![alt text](https://kurg.org/media/mlpack-www-2.jpg)

![alt text](https://kurg.org/media/mlpack-www-3.jpg)

gmanlan commented:
Hey @zoq: thanks for sharing. I like the ideas of adding NumFocus and Benchmark data, that surely adds value to the table and I couldn't find that information before, so I'm glad you bring it.
On the other side, it seems you have redesigned the website, and therefore we would need to "start again". I spent a couple of months working on this version (that we discussed over here: https://github.com/mlpack/mlpack/issues/1892) and unfortunately I won't have more time to re-do it again. If folks like the blue design better, would you have time and bandwidth to re-do it?

@rcurtin: I will hold the code changes from your review until we decide on this.

Cheers

rcurtin commented:
@gmanlan I think maybe @zoq's changes can be done in such a way that they're pretty minor.  Really it seems like just some CSS tweaks plus some tweaks to the header HTML on the front page.  (For instance, can the red/blue change be done with just some simple regex or a quick SCSS change?  I'm not sure.)  But if @gmanlan is short on time, we should maybe try to make the changes minimal here and then address things like adding the notebook interface / benchmarks / style changes in separate PRs?

From my opinion I see awesome ideas in both designs and I think the ideas are pretty complementary:

 * I really like the cards and flow of @gmanlan's design.
 * I like the red/black coloring of @gmanlan's design since that sticks with the logo colors.  Maybe the blue can be replaced with a subtle red tone or maybe even a gray tone in @zoq's mockup?  I don't have a strong opinion.
 * The SCSS improvements from @gmanlan are really nice, apparently I was not very good at using SCSS to its full power.
 * I like the notebook idea from @zoq, although we should be careful that we run significant amounts of code for every visitor to mlpack.org... we'll light some machines on fire like that.  Logs indicate that about 4k unique IPs visited mlpack.org yesterday.
 * I like the idea of adding benchmarks from @zoq's design, or maybe a card to benchmarks where we can have a more complete page or something?

Maybe we should proceed like this---once we can find some consensus on which parts of each design we should take and how we should fit them together, @gmanlan if you want to point out the comments I left that you'd like to take care of / leave for later, we can resolve the ones for later and open Github issues for them.  Then we can merge this, and @zoq could open issues/PRs for additional changes for other parts of the design.  In the past the mlpack.org repo has been pretty ignored (I've just committed various changes as needed), but it wouldn't be a bad idea to actually start tracking the things that we want to change.

Anyway, let me know what each of you think.  I'm excited that we are all seeing great ways to improve the website, and I think no matter what we come out with something that's better in basically every way.  (If you want to see how far we've come, here's a blast from the past: http://web.archive.org/web/20120607071412/http://www.mlpack.org/.  Another testament to the fact that I am not a designer. :))

gmanlan commented on _layouts/default-nav.html:
```diff
@@ -1,14 +1,15 @@
 <!DOCTYPE html>
 <html>
+<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.8.2/css/all.css" integrity="sha384-oS3vJWv+0UjzBfQzYUhtDYW+Pj2yciDJxpsK1OYPAYjqT085Qq/1cq5FLXAZQ7Ay" crossorigin="anonymous">
```
No need to keep a local copy, as it's served by CDN and includes integrity check.

gmanlan commented on _sass/typography.scss:
```diff
@@ -3,31 +3,33 @@
 // Based on the typographic scale: 12, 14, 16, 18, 21, 24, 36, 48, 60, 72.
 //
 
+@import url('https://fonts.googleapis.com/css?family=Roboto');
```
Same comment as before, this is distributed by CDNs and therefore would never be offline. Storing frozen copies of external resources in our repo is usually discouraged when possible. Worse case scenario, the next available font will takeover (which I guess is Arial or similar).

rcurtin replied:
Ah, ok.  I did not realize that. :)

gmanlan commented on _sass/layout.scss:
```diff
@@ -207,12 +206,13 @@ img {
   color: lighten($text-color, 35%);
   background-color: lighten($background-color, 35%);
 }
+*/
```
Done

gmanlan commented on community.md:
```diff
@@ -2,13 +2,13 @@
 ---
 
-## community
+<div class="page-title-header">Community</div>
```
I couldn't enforce the ## symbol to render the header that we need, therefore we use the custom class to format the title header, which is consistent across all the pages of the website.

rcurtin replied:
Oops, left the "TODO" comment there.  It was a placeholder for me to write a more coherent comment but it looks like you got the basic idea of the question I would have written anyway. :)  Anyway, sounds good to me, thanks for the explanation. :+1:

gmanlan commented on community.md:
```diff
@@ -20,14 +20,35 @@ page](gsoc.html).
+	    <p><a href="http://www.mlpack.org/blog/">Blog</a></p><p>for mlpack development updates and blog posts</p>
+	  </div>
+  </li>
+</ul>
```
Agree, and done. BTW, I couldn't find the chat services changes in the master community.md file, did you forget to push? I ended up copying and pasting from the live website anyways. All done.

rcurtin replied:
Yeah, like I said in another comment, I haven't pushed because I didn't do my homework yet to figure out how to push back to this repo without including the huge LFS objects. :(  I'll get it sorted...

gmanlan commented on _sass/custom.scss:
```diff
@@ -271,3 +260,297 @@ select.doxygen-version-select option
+  height: 150px;
+  text-align: left;
+  background-color: #ffffff;
+  cursor: pointer;
```
Good catch, fixed so cards are fully clickable now.

gmanlan commented on docs.md:
```diff
@@ -1,6 +1,37 @@
+	* [Method-specific Tutorials](doc/mlpack-3.1.1/cli_documentation.html#method-specific-tutorials)
+	* [Advanced Tutorials](doc/mlpack-3.1.1/cli_documentation.html#advanced-tutorials)
+	* [Policy Class Documentation](doc/mlpack-3.1.1/cli_documentation.html#policy-class-documentation)
+	* [Writing an mlpack binding](doc/mlpack-3.1.1/doxygen/iodoc.html)
```
I agree, although I'm still not sure about the categories. You are missing stuff such as "Build", which seems to be C++ specific but then "it's not" if you are planning to build the python bindings. The same blur zone happens with some of the "General" category entries. It seems like we don't have a criteria for writing/splitting the documentation, so sometimes it's driven by language, other times by feature and other times by a mix of both!
I think we will need to first work on re-shaping or re-organizing the documentation and then we can jump into cards. Let's plan to do it as an additional story once we have clearly determined the categories. In the meantime, I think the list is more maintainable if we need to quickly change/add something in the short-time (which I expect to happen).

rcurtin replied:
Yeah, totally agreed.  The categories *might* have made sense at one point in time but it's pretty muddy now (in fact, maybe some of those tutorials need to be thrown out or rewritten entirely).  We can definitely open additional issues for stuff like this.

gmanlan commented on getstarted.md:
```diff
@@ -0,0 +1,43 @@
+<h2>Python</h2>
+
+<h3> Get via pip or conda </h3>
+- <a href="doc/mlpack-3.1.0/doxygen/python_quickstart.html"> mlpack in Python quickstart guide </a>
```
Agree, but is that true for pip? I tried that last week and didn't work for me.

Also, regarding the package manager list, that was the original idea that @zoq started to work with, but I guess it requires additional effort to first ensure that packages are published/updated for each distribution. I don't own that master list, but it would be nice to start compiling it.

Again, as soon as we have that, adding it to the website is trivial!

rcurtin replied:
On Windows I haven't managed to make pip work yet for mlpack, but it would work for Linux and OS X.  I agree, once we get the list together adding it would be easy for sure. :+1:

gmanlan commented on getstarted.md:
```diff
@@ -0,0 +1,43 @@
+- <a href="doc/mlpack-3.1.0/doxygen/build_windows.html"> Installation through vcpkg </a>
+
+<h3> Build from source </h3>
+- <a href="doc/mlpack-3.1.0/doxygen/build_windows.html"> Windows build tutorial </a>
```
I agree, although the "simple card" that we have now doesn't work for the purpose (just incompatible CSS). I believe a "big/long card" version would be required, so the card can summarize the content in itself (instead of adding more clicks to the UX, which is not recommended).
Let's add an additional story to work on that separately, as I can't get it reworked today.

gmanlan commented on _sass/custom.scss:
```diff
@@ -271,3 +260,297 @@ select.doxygen-version-select option
+  
+  background: linear-gradient(45deg, rgba(255,175,189,.3), rgba(100,216,243,.3), rgba(234,236,198,.3), rgba(245,146,176,.3), rgba(52,219,216,.3)) 0 0 / 1000% no-repeat, url('{{ site.baseurl }}/../../img/banner2.png') 0 0 / cover no-repeat;
+  -webkit-animation: gradientAnimation 40s ease infinite;
+  animation: gradientAnimation 40s ease infinite;
```
Maybe we can compile a bunch of pictures (charts/diagrams) from different projects (i.e. neural networks, clustering, etc.) that were created using mlpack, and combine them with shadow gradient - something like this: https://d3js.org/ but obscured so it's seamlessly fused with the hero banner color/pattern.

gmanlan commented on questions.md:
```diff
@@ -28,8 +31,8 @@ to help with the bug report process.
-    See the [community](community.html#developers) page for more details and a
-    list.  mlpack is a community-led effort.
+    mlpack is a community-led effort. See the [community](community.html#developers) page for more details and a
+    list.
```
Done

gmanlan commented on index.md:
```diff
@@ -4,40 +4,86 @@ title: Home
+      <div class="card">
+	    <i class="fa fa-question-circle fa-lg card-icon"></i>
+          <h4>FAQ</h4>
+          <p>For those questions you usually have...</p>
```
Done

gmanlan commented on index.md:
```diff
@@ -4,40 +4,86 @@ title: Home
+      <div class="card">
+	    <i class="fa fa-book fa-lg card-icon"></i>
+		  <h4>Documentation</h4>
+		  <p>Take a look at the API, python bindings and more</p>
```
Done

gmanlan commented on index.md:
```diff
@@ -4,40 +4,86 @@ title: Home
+	</div>
+</div>
+
+<p class="intro-header"> Why mlpack </p>
```
Done

gmanlan commented:
@rcurtin and @zoq: I agree with the design change proposals and additions.
As you said, I believe we should tackle major changes with future issues/PRs and with more hands on deck. As far as I know, we all agreed on the initial design of the website and all the interested parties provided feedback when discussed months ago: https://github.com/mlpack/mlpack/issues/1892#issuecomment-499720428

If we introduce radical changes at this point, we will have been wasting months of (interrupted) work. :blush: Anyway, if somebody wants to take over, feel free.

Said that, I think I'm done with the code review changes. I resolved most of the comments but I'm leaving a couple of them opened so you can check/review or act on an AR.

I may have 1 extra day for final polishing, but I'm afraid I won't be available anymore for a while.

Looking forward to hearing from you.

rcurtin commented:
Blah, well, crap.  I force pushed the original repository (because I needed to ensure that it wasn't using LFS storage and didn't have large files) and then Github closed this PR.  That's not what I wanted.  Certainly during merge, I would have to probably cherry-pick everything.  So I can do that now.  @zoq @gmanlan let me know when each of you are done with the comments and I'll go ahead and cherry-pick to do the merge and deploy the new website.  We can still keep commenting on the closed PR (it seems we won't have any other choice... sorry about that).

zoq commented:
I'm happy with the current design, so whenever you are ready let's deploy the new website. Also, I like the idea to open separate PR's for further changes.

I guess, we can setup some sort of publish pipeline that makes it easier to deploy new changes?

rcurtin commented:
Yeah, definitely, I'll set up a job on ci.mlpack.org to operate just like the ensmallen website.

gmanlan commented:
@rcurtin, @zoq: I'm done with the changes from the comments. Just pushed one last commit so we should be ready to go (hopefully you can pull that even though the PR is closed). Please go ahead with the deployment and feel free to comment back here if you encounter issues. Thanks!

rcurtin commented:
@gmanlan awesome, thanks.  I'll cherry-pick the commits, do a test deployment, then deploy to mlpack.org, and then open a bunch of issues that we can handle going forwards that were mentioned in this PR.  Thanks so much for putting some love into the website!  I think it was much needed. :)

conradsnicta commented:
new website has link to old mlpack: https://github.com/mlpack/mlpack/issues/2109

rcurtin commented:
Alright, the initial deployment is done and I've worked out all the bugs.  The master branch here on Github is now the same as what's on my system.  I'll now open issues from the discussions we had here that we can address individually going forward. :+1:  Thanks again @gmanlan for the hard work here!  Really happy to have it deployed.

gmanlan commented:
@rcurtin: thanks for all your help. This is fantastic and I'm happy to see it live!
